tests/run/nogil_conditional.pyx

```diff
@@ -0,0 +1,242 @@
+# mode: run
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+
+def test(int x):
+    """
+    >>> test(0)
+    110
+    """
+    with nogil(True):
+        x = f_nogil(x)
+        with gil(True):
+            x = f_gil(x)
+    return x
+
+
+cdef int f_nogil(int x) nogil:
+    cdef int y
+    y = x + 10
+    return y
+
+
+cdef int f_gil(int x) with gil:
+    cdef int y
+    y = x + 100
+    return y
+
+
+cdef int with_gil_func() except 0 with gil:
+    raise Exception("error!")
+
+
+cdef int nogil_func() nogil except 0:
+    with_gil_func()
+
+
+def test_nogil_exception_propagation():
+    """
+    >>> test_nogil_exception_propagation()
+    Traceback (most recent call last):
+       ...
+    Exception: error!
+    """
+    with nogil:
+        with gil:
+            with nogil(True):
+                nogil_func()
+
+
+cdef int write_unraisable() nogil:
+    with gil:
+        raise ValueError()
+
+
+def test_unraisable():
+    """
+    >>> print(test_unraisable())  # doctest: +ELLIPSIS
+    ValueError
+    Exception...ignored...
+    """
+    import sys
+    old_stderr = sys.stderr
+    stderr = sys.stderr = StringIO()
+    try:
+        write_unraisable()
+    finally:
+        sys.stderr = old_stderr
+    return stderr.getvalue().strip()
+
+
+def test_nested():
+    """
+    >>> test_nested()
+    240
+    """
+    cdef int res = 0
+
+    with nogil(True):
+        res = f_nogil(res)
+        with gil(1 < 2):
+            res = f_gil(res)
+            with nogil:
+                res = f_nogil(res)
+
+        with gil:
+            res = f_gil(res)
+            with nogil(True):
+                res = f_nogil(res)
+            with nogil:
+                res = f_nogil(res)
+
+    return res
+
+
+def test_nested_condition_false():
+    """
+    >>> test_nested_condition_false()
+    220
+    """
+    cdef int res = 0
+
+    with gil(False):
+        res = f_gil(res)
+        with nogil(False):
+            res = f_gil(res)
+
+        with nogil(True):
+            res = f_nogil(res)
+            with gil(False):
+                res = f_nogil(res)
+
+    return res
+
+def test_try_finally():
+    """
+    >>> test_try_finally()
+    113
+    """
+    cdef int res = 0
+
+    try:
+        with nogil(True):
+            try:
+                res = f_nogil(res)
+                with gil(1 < 2):
+                    try:
+                        res = f_gil(res)
+                    finally:
+                        res += 1
+            finally:
+                res = res + 1
+    finally:
+        res += 1
+
+    return res
+
+
+ctypedef fused fused_with_object:
+    int
+    object
+
+
+def test_fused_object(fused_with_object x):
+    """
+    >>> test_fused_object[object]("spam")
+    456
+    >>> test_fused_object[int](1000)
+    1000
+    """
+    cdef int res = 0
+
+    if fused_with_object is object:
```

Comment:
I think this condition makes the test a bit boring, because it discards all nogil tests in the `int` case without even looking at them, so the two alternative cases are not actually tested. Maybe you could add a third fused types test that tries to actually use the GIL differently based on the type?

Reply:
When adding a test for different behavior based on the actual fused type I see that in these cases, the condition (e.g. `numeric is float`) is not constant during GilCheck which fails the test.

Should I add logic to check for these special kind of expressions (reuse the logic from `ReplaceFusedTypeChecks`) or should I change some previous transform so that these expressions become constant?

Reply:
It looks like a simple `.visit_GILStatNode()` in that transform that works like `.visit_IfStatNode()` should suffice.
(Actually, I wonder if it wouldn't have been best to make `RepFTCh` inherit from `ConstFold`, but that's not something to fix up right now.)

Reply:
> in that transform

You are referring to `ReplaceFusedTypeChecks`, right?

Reply:
Added it to `ReplaceFusedTypeChecks`, worked like a charm.